tests/test_dynamic_links.py

```diff
@@ -0,0 +1,212 @@
+        assert len(link_stats.event_stats) == 7
+        assert repr(link_stats.event_stats[6]) == repr(
+            dynamic_links.EventStats(
+                platform=dynamic_links.PLATFORM_IOS,
+                event=dynamic_links.EVENT_TYPE_REDIRECT,
+                count=124))
+
+
+    def test_get_stats_error(self, dltest):
+        dltest._instrument_dynamic_links(payload=MOCK_GET_STATS_RESPONSE,
+                                         status=500)
+        options = dynamic_links.StatOptions(duration_days=9)
+        with pytest.raises(requests.exceptions.HTTPError) as excinfo:
+            dynamic_links.get_link_stats(_MOCK_SHORT_URL, options, app=dltest.app)
+        assert excinfo.value.response.status_code == 500
+
+    @pytest.mark.parametrize('invalid_url', ['google.com'] + INVALID_STRINGS)
+    def test_get_stats_invalid_url(self, dltest, invalid_url):
+        options = dynamic_links.StatOptions(duration_days=9)
+        with pytest.raises(ValueError) as excinfo:
+            dynamic_links.get_link_stats(invalid_url, options, app=dltest.app)
+        assert 'Url must be a string and begin with "https://".' in str(excinfo.value)
+
+    @pytest.mark.parametrize('invalid_options', INVALID_STRINGS)
+    def test_get_stats_invalid_options(self, dltest, invalid_options):
+        with pytest.raises(ValueError) as excinfo:
+            dynamic_links.get_link_stats(_MOCK_SHORT_URL, invalid_options, app=dltest.app)
+        assert 'Options must be of type StatOptions.' in str(excinfo.value)
+
+    @pytest.mark.parametrize('invalid_duration', [0] + INVALID_NON_NEGATIVE_NUMS)
+    def test_get_stats_invalid_duration_days(self, dltest, invalid_duration):
+        options = dynamic_links.StatOptions(duration_days=invalid_duration)
+        with pytest.raises(ValueError) as excinfo:
+            dynamic_links.get_link_stats(_MOCK_SHORT_URL, options, app=dltest.app)
+        assert 'duration_days' in str(excinfo.value)
+        assert 'must be positive int' in str(excinfo.value)
+
+
+class TestEventStats(object):
+    @pytest.mark.parametrize('platform', dynamic_links.EventStats._platforms.keys())
+    def test_valid_platform_values(self, platform):
+        e_s = dynamic_links.EventStats(
+            platform=dynamic_links.EventStats._platforms[platform],
+            event=dynamic_links.EVENT_TYPE_CLICK,
+            count=1)
+        assert e_s.platform == dynamic_links.EventStats._platforms[platform]
+
+    @pytest.mark.parametrize('arg', INVALID_STRINGS + ['unrecognized'])
+    def test_invalid_platform_values(self, arg):
+        with pytest.raises(ValueError) as excinfo:
+            dynamic_links.EventStats(
+                platform=arg,
+                event=dynamic_links.EVENT_TYPE_CLICK,
+                count=1)
+        assert 'not recognized' in str(excinfo.value)
+
+    @pytest.mark.parametrize('arg', dynamic_links.EventStats._platforms.keys())
+    def test_raw_platform_values_invalid(self, arg):
```

Comment:
I feel like a lot of this could have been avoided or at least simplified, if all the error checking happened in the constructor.

Reply:
Moved checks to constructor, leaving this for coverage.